Add normative wording and use set-cookie value format (#20)
w3c/correlation-context: correlation_context/HTTP_HEADER_FORMAT.md

The header contains a `list` of key-value pairs with optional properties attached.

This section uses the Augmented Backus-Naur Form (ABNF) notation of [[!RFC5234]], including the DIGIT rule in <a data-cite='!RFC5234#appendix-B.1'>appendix B.1 for RFC5234</a>. It also includes the `OWS` rule from <a data-cite='!RFC7230#whitespace'>RFC7230 section 3.2.3</a>.

## Definition

```
list        = list-member 0*179( OWS "," OWS list-member )
list-member = key OWS "=" OWS value *( OWS ";" OWS property )
property    = key OWS "=" OWS value
property    = key OWS
key         = <token, defined in [[RFC2616], Section 2.2](https://tools.ietf.org/html/rfc2616#section-2.2)>
value       = %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E
              ; US-ASCII characters excluding CTLs,
              ; whitespace DQUOTE, comma, semicolon,
              ; and backslash
OWS         = <Optional white space, as defined in [[RFC7230], Section 3.2.3.](https://tools.ietf.org/html/rfc7230#section-3.2.3)>
```

## Limits
1. Maximum number of name-value pairs: `180`.
2. Maximum number of bytes per a single name-value pair: `4096`.
3. Maximum total length of all name-value pairs: `8192`.

## Example

> `key1=value1[;property1;property2], key2 = value2[;property3]`

Reviewer: Can we include a property with a value? like `property4=propertyvalue` or similar?

Author: fixed - the example also included [] which is wrong

# Correlation Context HTTP Header Format

> The correlation context header is used to propagate user-supplied key-value pairs through a trace.

Reviewer 2: There still may be value in comparing this header to traceparent, so that readers aren't left confused about why both exist

# Correlation Context HTTP Header Format

> The correlation context header is used to propagate key-value context properties through a trace.

Reviewer 3: ```suggestion
The correlation context header is used to propagate user-supplied key-value pairs through a trace.
```

Reviewer 3: NB: I think we should not be referring to "trace" in this spec.

A received header MAY be altered to change or add information and it MUST be passed on to all downstream request.

Multiple correlation context headers are allowed. Values can be combined in a single header according to  [RFC 7230](https://tools.ietf.org/html/rfc7230#page-24).

*See [rationale document](HTTP_HEADER_FORMAT_RATIONALE.md) for details of decisions made for this format.*

# Header Name

Header name: `Correlation-Context`

@TODO: Agree on the header name
@TODO: Add more information on casing once we agree on the final header name

# Header Content

> This section uses the Augmented Backus-Naur Form (ABNF) notation of [[!RFC5234]], including the DIGIT rule in <a data-cite='!RFC5234#appendix-B.1'>appendix B.1 for RFC5234</a>. It also includes the `OWS` rule from <a data-cite='!RFC7230#whitespace'>RFC7230 section 3.2.3</a>.

Reviewer 3: ```suggestion
This section uses the Augmented Backus-Naur Form (ABNF) notation of [[!RFC5234]], including the DIGIT rule in <a data-cite='!RFC5234#appendix-B.1'>appendix B.1 for RFC5234</a>. It also includes the `OWS` rule (optional whitespace) from <a data-cite='!RFC7230#whitespace'>RFC7230 section 3.2.3</a>.
```

this simple clarification makes it unnecessary to look up the other RFC (whereas DIGIT is self-explanatory)

## Definition

```
list        = list-member 0*179( OWS "," OWS list-member )
list-member = key OWS "=" OWS value *( OWS ";" OWS property )
property    = key OWS "=" OWS value
property    = key OWS
key         = <token, defined in [[RFC2616], Section 2.2](https://tools.ietf.org/html/rfc2616#section-2.2)>
value       = %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E
              ; US-ASCII characters excluding CTLs,

>               ; whitespace DQUOTE, comma, semicolon,

Reviewer 3: did you mean to have a comma?
```suggestion
              ; whitespace, DQUOTE, comma, semicolon,
```

Author: This was taken directly from RFC2616 but, yes it is missing a comma

>               ; and backslash

Reviewer 3: What is the rationale for excluding DQUOTE and backslash? Is that taken from set-cookie?

Reviewer: Yes that is a direct copy from the RFC https://tools.ietf.org/html/rfc6265#section-4.1

Author: Yes, this was for the sake of consistency with `set-cookie`. Also, the backslash is often used for quoting. While this isn't the case for `urlencode`, that might be the reason.

OWS         = <Optional white space, as defined in [[RFC7230], Section 3.2.3.](https://tools.ietf.org/html/rfc7230#section-3.2.3)>
```

## Limits
1. Maximum number of name-value pairs: `180`.
2. Maximum number of bytes per a single name-value pair: `4096`.
3. Maximum total length of all name-value pairs: `8192`.

## Example

> `key1=value1;property1;property2, key2 = value2; property3, key3=value3;propertyKey=propertyValue`

Reviewer 3: ```suggestion
`key1=value1;property1;property2, key2 = value2, key3=value3; propertyKey=propertyValue`
```

this way we have an example with no properties

### list
List of key-value pairs with optional properties attached.
It can not be guaranteed that keys are unique.

Consumers MUST be able to handle duplicate keys while producers SHOULD try to dedupe the list.

### key
ASCII string according to the `token` format, defined in [[RFC2616], Section 2.2](https://tools.ietf.org/html/rfc2616#section-2.2). Leading and trailing OWS is allowed but MUST be trimmed when converting the header into a data structure.

### value

> The value contains an URL encoded UTF-8 string. Leading and trailing OWS is allowed but MUST be trimmed when converting the header into a data structure.

Reviewer 3: > URL encoded ... string

I though we agreed not to require URL encoding, leaving it to producers/consumers of a given key to agree on representation that satisfies the protocol restrictions.

Reviewer: We were leaning that way when you disconnected, but after talking about it more it seems there is not strong agreement either way. Mostly the reason for the encoding is that the values will somehow have to be encoded anyway so prescribing an encoding scheme makes it easier for e.g. APM vendors to read and display to users keys they may not "know" about.

Reviewer 3: nvm

Reviewer 3: Fair enough, I don't have a strong objection. It might be a bit weird when used with non-HTTP transports, but then we still do need to escape the commas, so url encoding is ok.

### property

> Properties MAY be appended to values and contain a keys & key-value pairs `;` delimited list `;k1=v1;k2;k3=v3`. The semantic of such properties is opaque to this specification.

Reviewer 3: ```suggestion
Additional metadata MAY be appended to values in the form of property set, represented as semi-colon `;` delimited list of keys and/or key-value pairs, e.g. `;k1=v1;k2;k3=v3`. The semantic of such properties is opaque to this specification.
```

## Example
`key1=value1[;property1;property2], key2 = value2[;property3]`

### list
List of key-value pairs with optional properties attached.
It can not be guaranteed that keys are unique.

> Consumers MUST be able to handle duplicate keys while producers SHOULD try to dedupe the list.

Reviewer: ```suggestion
Consumers MUST be able to handle duplicate keys while producers SHOULD try to deduplicate the list.
```

Reviewer 3: I think we should prescribe how deduping should be done, e.g. last entry wins
